src/domains/hoareDomain.ml

```diff
@@ -194,7 +193,7 @@ struct
   let narrow = product_bot (fun x y -> if B.leq y x then B.narrow x y else x)
 
   let add x a = if mem x a then a else add x a (* special mem! *)
-  let remove x a = unsupported "Set.remove"
+  let remove x a = filter (fun y -> not (B.leq y x)) a
```

Comment:
I don't think we should expose this functionality as `remove` as it is very different from the normal `remove` on a set.

Consider e.g. a set from the chain lattice over {a,b,c}. The element {a,b} is represented by {b}, removing {b} now yields {} here, the normal set remove on the other hand would yield {a}, which is larger in the lattice order.
So this "remove" is not a sound abstraction of the "remove" of sets which one would expect.

Reply:
This is necessary for `AddressDomain` to work though. `AD.remove` and `AD.remove` are used for some crucial logic, which is why I had to implement those for `SetDomain.Joined` as well (otherwise the reachability loop in base never terminates because it cannot remove anything from the workqueue set, or ever refine any values out of address sets).

`HoarePO` also did it in this way and that's where the behavior is from. I agree these removals can be problematic, so I'll add comments to them warning about this.

Reply:
To me it looks like `AD.remove` is only called with very specific values `null_ptr` and `unknown_ptr` for which this is fine. So I think we should maybe not expose this functionality everywhere, or at least name it something like `remove_all_smaller`.

Reply:
There's still the more general `diff`, which is also used with non-singleton pointer variants.

The annoying thing about having a separate name for this is that it requires a new module type for such sets and thus duplication of most of the set domains hierarchy we have. Only `SetDomain.Make` would be spared from this, because there `remove_all_smaller` would coincide with `remove`. And that's simply because `SetDomain.Make` (whose elements are `Printable.S`) is just a specific instance of set abstractions (whose elements are `Lattice.S`) where the element lattice is flat/fake, thus all equal and smaller non-bot elements to be removed are just the element itself.

Actually this sort of `remove` is the only meaningful one for set abstractions because in all other non-concrete abstractions you can never remove anything, since you cannot be sure (in general) that the elements of the set abstraction are singletons. That's only the case if the elements belong to a flat/fake lattice, which is implicitly the case for `SetDomain.Make` and in `AddressDomain` elements like `NullPtr` and `UnknownPtr`.

Declaring `remove` to mean this kind of _strong_ removal for all set abstractions wouldn't change anything for `SetDomain.Make`. For `SetDomain.ToppedSet` `diff` actually does a _strong_ removal (the first case):
https://github.com/goblint/analyzer/blob/13da7aca2966d9454a527a382e938dbbbe38e0d2/src/domains/setDomain.ml#L209-L213
If it were a _weak_ removal, then the first case would have to return `x` instead: since an unknown set `y` of elements might not cover all the elements of `x`.

Reply:
We should at least add a comment to this here, to indicate that it's usage is dangerous!

Reply:
I documented this both in the `SetDomain` interface and the specific implementations.